gio/src/subclass/file_enumerator.rs

```diff
@@ -0,0 +1,597 @@
+    let imp = instance.imp();
+    let cancellable = Option::<Cancellable>::from_glib_none(cancellable);
+    let callback = callback.map(move |callback| {
+        move |source_obj: Option<&glib::Object>, res: &AsyncResult| unsafe {
+            callback(source_obj.to_glib_none().0, res.to_glib_none().0, user_data)
+        }
+    });
+
+    imp.close_async(from_glib(io_priority), cancellable.as_ref(), callback);
+}
+
+unsafe extern "C" fn close_finish<T: FileEnumeratorImpl>(
+    enumerator: *mut ffi::GFileEnumerator,
+    result: *mut ffi::GAsyncResult,
+    error: *mut *mut glib::ffi::GError,
+) -> glib::ffi::gboolean {
+    let instance = &*(enumerator as *mut T::Instance);
+    let imp = instance.imp();
+    let result = Option::<AsyncResult>::from_glib_none(result);
+
+    let res = imp.close_finish(result.as_ref());
+
+    match res {
+        Ok(_) => true.into_glib(),
+        Err(err) => {
+            if !error.is_null() {
+                *error = err.to_glib_full()
+            }
+            false.into_glib()
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::path::PathBuf;
+
+    use futures_channel::oneshot;
+
+    use super::*;
+    use crate::{prelude::*, IOErrorEnum};
+
+    mod imp {
+        use std::sync::RwLock;
+
+        use super::*;
+
+        #[derive(Default)]
+        pub struct MyFileEnumerator {
+            pub index: RwLock<i8>,
+        }
+
+        #[glib::object_subclass]
+        impl ObjectSubclass for MyFileEnumerator {
+            const NAME: &'static str = "MyFileEnumerator";
+            type Type = super::MyFileEnumerator;
+            type ParentType = FileEnumerator;
+        }
+
+        impl ObjectImpl for MyFileEnumerator {}
+
+        impl FileEnumeratorImpl for MyFileEnumerator {
+            fn next_file(
+                &self,
+                _cancellable: Option<&Cancellable>,
+            ) -> Result<Option<FileInfo>, glib::Error> {
+                let mut index = self.index.write().unwrap();
+                match *index {
+                    -1 => Err(glib::Error::new(IOErrorEnum::Closed, "Closed")),
+                    0..10 => {
+                        let fileinfo = glib::Object::builder::<FileInfo>().build();
+                        fileinfo.set_name(format!("file{}", *index));
+                        *index = *index + 1;
+                        Ok(Some(fileinfo))
+                    }
+                    _ => Ok(None),
+                }
+            }
+
+            fn close(&self, _cancellable: Option<&Cancellable>) -> Result<(), glib::Error> {
+                let mut index = self.index.write().unwrap();
+                *index = -1;
+                Ok(())
+            }
+        }
+    }
+
+    glib::wrapper! {
+        pub struct MyFileEnumerator(ObjectSubclass<imp::MyFileEnumerator>) @extends FileEnumerator;
+    }
+
+    unsafe impl Send for MyFileEnumerator {}
+    unsafe impl Sync for MyFileEnumerator {}
```

Comment:
These are implemented automatically if possible. Are they not, and why are they required here?